evm-fungible-token/src/lib.rs

```diff
@@ -0,0 +1,351 @@
+            event.amount
+        );
+
+        assert_eq!(
+            event.eth_custodian_address,
+            self.eth_custodian_address,
+            "Event's address {} does not match custodian address {}",
+            hex::encode(&event.eth_custodian_address),
+            hex::encode(&self.eth_custodian_address),
+        );
+        let proof_1 = proof.clone();
+        let account_id = env::current_account_id();
+        let prepaid_gas = env::prepaid_gas();
+        log!(
+            "Deposit verify_log_entry for prover: {:?}",
+            self.prover_account,
+        );
+        let promise0 = env::promise_create(
+            self.prover_account.clone(),
+            b"verify_log_entry",
+            json!({
+                "log_index": proof.log_index,
+                "log_entry_data": proof.log_entry_data,
+                "receipt_index": proof.receipt_index,
+                "receipt_data": proof.receipt_data,
+                "header_data": proof.header_data,
+                "proof": proof.proof,
+                "skip_bridge_call": false,
+            })
+            .to_string()
+            .as_bytes(),
+            NO_DEPOSIT,
+            prepaid_gas / 4,
+        );
+        let promise1 = env::promise_then(
+            promise0,
+            account_id,
+            b"finish_deposit",
+            json!({
+                "new_owner_id": event.recipient,
+                "amount": event.amount,
+                "proof": proof_1,
+            })
+            .to_string()
+            .as_bytes(),
+            NO_DEPOSIT,
+            prepaid_gas / 4,
+        );
+        env::promise_return(promise1);
+    }
+
+    /// Finish depositing once the proof was successfully validated.
+    /// Can only be called by the contract itself.
+    #[private]
+    pub fn finish_deposit(&mut self, new_owner_id: AccountId, amount: U128, proof: Proof) {
+        log!(
+            "finish_deposit - Promise results: {:?}",
+            env::promise_results_count()
+        );
+        log!("Amount: {:?}", amount);
+        assert_eq!(env::promise_results_count(), 1);
+        let data0: Vec<u8> = match env::promise_result(0) {
+            PromiseResult::Successful(x) => x,
+            _ => panic!("Promise with index 0 failed"),
+        };
+        let verification_success: bool = serde_json::from_slice(&data0).unwrap();
+        assert!(verification_success, "Failed to verify the proof");
+        self.record_proof(&proof.get_key());
+
+        self.mint(new_owner_id, amount.into());
+    }
+
+    /// Mint Fungible Token for account
+    /// TODO: should be related to NEP-145
+    #[private]
+    fn mint(&mut self, owner_id: AccountId, amount: Balance) {
+        log!("Mint {:?} tokens for: {:?}", amount, owner_id);
+        if self.token.accounts.get(&owner_id).is_none() {
+            self.token.accounts.insert(&owner_id, &amount);
+        } else {
+            self.token.internal_deposit(&owner_id, amount);
+        }
```

Comment:
I think mint should also update total_supply?

Reply:
> I think mint should also update total_supply?

@djsatok we do not discuss that thing. What do you think?

Reply:
@ailisp , the total supply is updated in `internal_deposit`, this piece of the code was agreed to be changed into
```
if self.token.accounts.get(&owner_id).is_none() {
  self.token.accounts.insert(&owner_id, 0);
} 
self.token.internal_deposit(&owner_id, amount);
```

Also, @mrLSD, please take a look at the creation of the FT using the committed standard: https://github.com/near/near-sdk-rs/blob/master/examples/fungible-token/ft/src/lib.rs

The issue with attaching a deposit with the deposit transaction is yet to be solved.